### DISCUSSION TITLE
LFE and hygenic macros?

### DISCUSSION THREAD

**Discussion by AeroNotix** [lfe/lfe#455] [Q&A] [answered]
Hi,

After an interesting talk with @rvirding in #erlang, it's clear that LFE does not have the utilities to create hygenic macros.

Take the following situation (Common Lisp):

``` lisp
(defmacro for (var start stop &body body)
 `(progn
    (format t "stopping at: ~d" ,stop)
    (do
    ((,var ,start (1+, var)))
    ((> ,var ,stop))
    ,@body)))
```

This works fine when we call it as such:

``` lisp
(for i 1 5
  (print i))
;;
stopping at: 5
1 
2 
3 
4 
5 
NIL
```

However, calling this macro with a call to another side-effecting function yields strange results:

``` lisp
CL-USER> (for i 1 (random 10) (print i))
stopping at: 9
1 
NIL
CL-USER> (for i 1 (random 10) (print i))
stopping at: 3
1 
2 
NIL
CL-USER> (for i 1 (random 10) (print i))
stopping at: 7
1 
2 
NIL
CL-USER> (for i 1 (random 10) (print i))
stopping at: 1
1 
2 
NIL
CL-USER> (for i 1 (random 10) (print i))
stopping at: 8
1 
NIL
CL-USER> (for i 1 (random 10) (print i))
stopping at: 7
1 
2 
NIL
CL-USER> (for i 1 (random 10) (print i))
stopping at: 5
1 
2 
NIL
CL-USER> (for i 1 (random 10) (print i))
stopping at: 2
1 
2 
3 
4 
NIL
CL-USER> (for i 1 (random 10) (print i))
stopping at: 0
NIL
CL-USER> (for i 1 (random 10) (print i))
stopping at: 2
1 
2 
3 
4 
5 
```

Which is _most definitely_ incorrect.

Ok, so how about we bind the stop call _before_ making the do block? Okay!

``` lisp
(defmacro for (var start stop &body body)
 `(let ((real-stop ,stop))
    (progn
      (format t "stopping at: ~d" real-stop)
      (do
       ((,var ,start (1+, var)))
       ((> ,var real-stop))
        ,@body))))
```

Yields:

``` lisp
CL-USER> (for i 1 (random 10) (print i))
stopping at: 6
1 
2 
3 
4 
5 
6 
NIL
CL-USER> (for i 1 (random 10) (print i))
stopping at: 3
1 
2 
3 
NIL
CL-USER> (for i 1 (random 10) (print i))
stopping at: 2
1 
2 
NIL
CL-USER> (for i 1 (random 10) (print i))
stopping at: 9
1 
2 
3 
4 
5 
6 
7 
8 
9 
NIL
```

Which is what we want! But wait...

What if the body of the macro was used as such:

``` lisp
(for i 1 (random 10) (setq real-stop :banana))
```

Yields (SBCL):

```
The value :BANANA is not of type REAL.
   [Condition of type TYPE-ERROR]

Restarts:
 0: [RETRY] Retry SLIME REPL evaluation request.
 1: [*ABORT] Return to SLIME's top level.
 2: [ABORT] Abort thread (#<THREAD "repl-thread" RUNNING {10036D8063}>)

Backtrace:
  0: ((LAMBDA ()))
  1: (SB-INT:SIMPLE-EVAL-IN-LEXENV (FOR I 1 (RANDOM 10) (SETQ REAL-STOP :BANANA)) #<NULL-LEXENV>)
  2: (EVAL (FOR I 1 (RANDOM 10) (SETQ REAL-STOP :BANANA)))
 --more--
```

Whoooops.

What about the wholly more insidious:

``` lisp
(for i 1 (random 10) (setq real-stop (random 100000)))
```

Which means the iteration stops on a random basis entirely.

Lisps get around this usually by using a call to `(gensym)` which creates a symbol which is _guaranteed_ to be unique, this means we can bind forms to a generated symbol thereby ensuring that the form is evaluated once and that the symbol will not clash with any captured body.

@rvirding mentioned to me that `(gensym)` would create atoms in the atom table. Number one, I do not understand fully how LFE is implemented. However, I don't understand why atoms are needed in the creation of a Variable label? You would just need to generate a variable name which can be used in the body.

Am I missing some information here?

**Comment by nox** [comment]
I don't think it's bad for `(gensym)` to generate new atoms, as it's a compile-time thing. But that being said, I remember that Core's inliner uses integers for the variables it emits, maybe you could do that too.

**Comment by rvirding** [comment, marked as answer]
Some comments about symbols and atoms.

Two of the goals I have for LFE is to make it as efficient as erlang and to make it completely compatible with erlang and OTP. These goals interact with each other. One result of this is that LFE can never be a CL or a scheme, there are too many properties of erlang which make this impossible, for example immutable data, its module system and handling of function arguments. And its data types.

To keep the homoiconicity of lisp I had to use the erlang datatypes directly and the closest to a CL symbol is the atom. It is different in that it has no value (except itself), function binding or properties and atoms are always interned in the atom table. This also means that it is impossible to make a unique atom, and that you can fill the atom table with "temporary" atoms and crash the systems. The macro expander reuses the same names for generated variables each time.

That's why I haven't implemented (gensym). Also it is not that critical in LFE as variables are scoped and there is no setq, or setf. You can reuse the same temporary variables and the worst that can happen is shadowing.

A possible solution would be for the compiler to allow some structure to represent temporary variables. This would be easy to implement but not safe but with this limitation it could give us hygenic macros.

It would not be possible to use integers as variables as the macro expanding is done in lisp while the Core comes after the code has been compiled. And Core variables are tagged structures.

**Comment by AeroNotix** [comment]
(gensym) is not just about mutable variables, it's also about in-order execution of expressions and the multiple evaluation of expressions.

Consider the following:

``` lisp
%% Forgive me if this is not the exact LFE syntax, I hope the idea shines through.
(defmacro error-ok (expr)
  `(case ((lambda () ,@expr))
     (tuple ok value
            (ok value))
     (tuple error value
            (error value))))
```

What would happen here if `value` was already in-scope? You would pattern match against that value, thus meaning that you _cannot_ introduce any lexical variables with macros.

The above is caused by macros _expanding into the body of a form_.

One way around this in typical Lisps (other than gensym!) is to evaluate the expressions inside a lambda and pass in the expression, but since LFE is Erlang and it pattern matches, the lambda expression will be a lexical closure and it'll pattern match against variables, rather than bringing in a new lexical binding, afaict.

**Comment by rvirding** [comment]
It is not quite that bad. LFE does share erlang's lack of variable scoping so the case creates a new scope and variables in the patterns are new local variables in side the case clauses. So the variable value will be a new variable in each clause which will be the one referred in their body.

This means that you explicitly have to add an explicit equality guard test when you do want to test against an existing. Also multiple occurrence of a variable in a pattern is illegal and you have to use different names with an explicit equality test.

**Comment by arpunk** [comment]
I'm bumping this in order to clarify and update the current LFE documentation to get it ready for the release.

@rvirding, @AeroNotix  any final words on this?
